keeper/src/main/scala/zio/keeper/membership/swim/protocols/FailureDetection.scala

```diff
@@ -0,0 +1,179 @@
+package zio.keeper.membership.swim.protocols
+
+import upickle.default._
+import zio.duration._
+import zio.keeper.membership.swim.Nodes.NodeState
+import zio.keeper.membership.swim.{ Message, Nodes, Protocol }
+import zio.keeper.membership.{ ByteCodec, NodeAddress, TaggedCodec }
+import zio.logging.Logging.Logging
+import zio.logging._
+import zio.stm.TMap
+import zio.stream.ZStream
+import zio.{ Ref, Schedule, UIO, ZIO, keeper }
+
+sealed trait FailureDetection
+
+object FailureDetection {
+
+  implicit def tagged(
+    implicit
+    c1: ByteCodec[Ack],
+    c2: ByteCodec[Ping],
+    c3: ByteCodec[PingReq],
+    c4: ByteCodec[Nack]
+  ): TaggedCodec[FailureDetection] =
+    TaggedCodec.instance(
+      {
+        case _: Ack     => 10
+        case _: Ping    => 11
+        case _: PingReq => 12
+        case _: Nack    => 13
+      }, {
+        case 10 => c1.asInstanceOf[ByteCodec[FailureDetection]]
+        case 11 => c2.asInstanceOf[ByteCodec[FailureDetection]]
+        case 12 => c3.asInstanceOf[ByteCodec[FailureDetection]]
+        case 13 => c4.asInstanceOf[ByteCodec[FailureDetection]]
+      }
+    )
+
+  final case class Ack(conversation: Long) extends FailureDetection
+
+  object Ack {
+
+    implicit val codec: ByteCodec[Ack] =
+      ByteCodec.fromReadWriter(macroRW[Ack])
+
+  }
+
+  final case class Nack(conversation: Long) extends FailureDetection
+
+  object Nack {
+
+    implicit val codec: ByteCodec[Nack] =
+      ByteCodec.fromReadWriter(macroRW[Nack])
+
+  }
+
+  final case class Ping(ackConversation: Long) extends FailureDetection
+
+  object Ping {
+
+    implicit val codec: ByteCodec[Ping] =
+      ByteCodec.fromReadWriter(macroRW[Ping])
+  }
+
+  final case class PingReq(target: NodeAddress, ackConversation: Long) extends FailureDetection
+
+  object PingReq {
+
+    implicit val codec: ByteCodec[PingReq] =
+      ByteCodec.fromReadWriter(macroRW[PingReq])
+  }
+
+  private case class _Ack(onBehalf: Option[(NodeAddress, Long)])
+
+  def protocol(nodes: Nodes, protocolPeriod: Duration, protocolTimeout: Duration) =
+    for {
+      acks  <- TMap.empty[Long, _Ack].commit
```

Comment:
I'd rename this to `pendingAcks` or `heartbeats`.

Reply:
and the winner is ```pendingAcks```